### PR TITLE
fix: Synchronization lock handling in Step/DAG Template level

### PR DESCRIPTION
**Pull request by sarabala1979**
Checklist:

* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo-workflows/blob/master/USERS.md).

### CODE REVIEW

File: workflow/controller/steps.go [hunk 1]
```diff
@@ -94,11 +94,7 @@ func (woc *wfOperationCtx) executeSteps(ctx context.Context, nodeName string, tm
 
 		sgNode := woc.executeStepGroup(ctx, stepGroup.Steps, sgNodeName, &stepsCtx)
 
-		if sgNode.Fulfilled() {
-			if tmpl.Synchronization != nil {
-				woc.controller.syncManager.Release(woc.wf, node.ID, tmpl.Synchronization)
-			}
-		} else {
+		if !sgNode.Fulfilled() {
```
Comment (sarabala1979) [inline workflow/controller/steps.go:97]: This is a bug. Code is releasing parent lock if child completes

File: workflow/controller/operator.go [hunk 2]
```diff
@@ -1785,6 +1785,12 @@ func (woc *wfOperationCtx) executeTemplate(ctx context.Context, nodeName string,
 		}
 	}
 
+	if node.Fulfilled() {
```
Comment (sarabala1979) [inline workflow/controller/operator.go:1788]: This is the right place to check the node status and release the lock.

File: examples/steps.yaml [hunk 3]
```diff
@@ -8,6 +8,11 @@ spec:
   entrypoint: hello-hello-hello
   templates:
   - name: hello-hello-hello
+    synchronization:
```
Comment (alexec) [inline examples/steps.yaml:11]: can we have the example in a different file?

File: workflow/controller/operator.go [hunk 4]
```diff
@@ -1785,6 +1785,12 @@ func (woc *wfOperationCtx) executeTemplate(ctx context.Context, nodeName string,
...
+		if processedTmpl.Synchronization != nil {
```
Comment (alexec) [inline workflow/controller/operator.go:1789]: I think you could merge the two if expressions into one statement

Comment (alexec) [reply workflow/controller/operator.go:1789]: what about adding these two checks to `.Release` - i.e. have it refuse to release if the node is unfulfilled or the synchronization is nil? the less code we add to woc, the better

File: workflow/controller/operator_concurrency_test.go [hunk 5]
```diff
@@ -486,3 +487,331 @@ func TestSynchronizationWithRetry(t *testing.T) {
 
 	})
 }
+
+const StepWithSync = `
```
Comment (alexec) [inline workflow/controller/operator_concurrency_test.go:491]: these examples are a bit long for me to understand - could you please simplify them are some how make them easier to understand? maybe add comments?

**Review by alexec — COMMENTED** [review]
Good work. I love how thorough your tests are.

Can I make an ask?

I see this logged a lot:

```
 controller | time="2021-02-11T08:36:08.398Z" level=info msg="Check the workflow existence"
```

Could we either:

1. Make this DEBUG?
2. Add some fields to it to give it a bit more context? Currently I need to look at the source code to understand it. Most users will not do that.

**Review by alexec — APPROVED** [review]

**Pull request merged**